src/analyses/ai_storage.h

```diff
@@ -0,0 +1,298 @@
+/// create and initialise the domains as needed.
+
+#ifndef CPROVER_ANALYSES_AI_STORAGE_H
+#define CPROVER_ANALYSES_AI_STORAGE_H
+
+#include <goto-programs/goto_program.h>
+
+#include "ai_domain.h"
+#include "ai_history.h"
+
+/// This is the basic interface for storing domains.
+/// The abstract interpreters should use this interface by default.
+class ai_storage_baset
+{
+public:
+  ai_storage_baset()
+  {
+  }
+
+  virtual ~ai_storage_baset()
+  {
+  }
+
+  typedef ai_domain_baset statet;
+  typedef std::shared_ptr<statet> state_ptrt;
+  typedef std::shared_ptr<const statet> cstate_ptrt;
+  typedef ai_history_baset tracet;
+  typedef ai_history_baset::trace_ptrt trace_ptrt;
+  typedef ai_history_baset::trace_sett trace_sett;
+  typedef std::shared_ptr<trace_sett> trace_set_ptrt;
+  typedef std::shared_ptr<const trace_sett> ctrace_set_ptrt;
+  typedef goto_programt::const_targett locationt;
+
+  /// Returns all of the histories that have reached
+  /// the start of the instruction.
+  virtual ctrace_set_ptrt abstract_traces_before(locationt l) const = 0;
+
+  /// Non-modifying access to the stored domains,
+  /// used in the ai_baset public interface.
+  /// In the case of un-analysed locals this should create a domain
+  /// The history version is the primary version, the location one may
+  /// simply join all of the histories that reach the given location
+  virtual cstate_ptrt abstract_state_before(
+    trace_ptrt p,
+    const ai_domain_factory_baset &fac) const = 0;
+
+  virtual cstate_ptrt abstract_state_before(
+    locationt l,
+    const ai_domain_factory_baset &fac) const = 0;
+
+  /// Look up the analysis state for a given history,
+  /// instantiating a new domain if required.
+  virtual statet &
+  get_state(trace_ptrt p, const ai_domain_factory_baset &fac) = 0;
+
+  /// Reset the abstract state
+  virtual void clear()
+  {
+    return;
+  }
+
+  /// Notifies the storage that the user will not need the domain object(s)
+  /// for this location.  After this has been called abstract_state_before may
+  /// return an over-approximation of the value and get_state may give an
+  /// under-approximation (forcing recomputation).
+  /// If there are multiple histories that reach this location all will be
+  /// affected
+  virtual void prune(locationt l)
+  {
+    return;
+  }
+};
+
+// There are a number of options for how to store the history objects.
+// This implements a simple one.  It is not in ai_storage_baset so that
+// storage implementations can implement their own, more specific, approaches
+class trace_map_storaget : public ai_storage_baset
+{
+protected:
+  typedef std::map<locationt, trace_set_ptrt> trace_mapt;
+  trace_mapt trace_map;
+
+  // This retains one part of a shared_ptr to the history object
+  void register_trace(trace_ptrt p)
+  {
+    // Save the trace_ptrt
+    trace_mapt::iterator it = trace_map.find(p->current_location());
+    if(it == trace_map.end())
+    {
+      trace_set_ptrt s(new trace_sett());
```

Comment:
Maybe... trace_sett is a typedef around standard library objects so it would require quite a bit of code for simplifying two calls.  If it were already a custom object then, hell yeah, but as it is I am not sure it is a net saving.